### ISSUE TITLE
mm2-fast output is not identical over various compilations

### ISSUE THREAD

**Issue by ASLeonard** [lh3/minimap2#793]
Hi,
This may be more for @SaurabhKalikar.

I was testing the mm2-fast branch, and the speed is **fast**. I've seen between 3-7x speedup depending on the compiler options when mapping about 15x hifi data to a reference.

However, I noticed when comparing the output of using only the lhash and using the full AVX512 option, and they are close but not the same, while mm2-fast claims it should be the `exact same output` as minimap2, so should be consistent among its own.

With
```
Using LISA hash with sequential last-mile search.
Using default chaining.
Using default SSE-vectorized alignment.
---
Number of mapped sequences: 1825625
Number of primary alignments: 2381249
Number of secondary alignments: 4397302

Number of bases in mapped sequences: 36042685390
Number of mapped bases: 35461391419
```

and
```
Using LISA hash with AVX512-vectorized last-mile search.
Using AVX512-vectorized chaining.
Using AVX512-vectorized alignment.
---
Number of mapped sequences: 1825625 [same]
Number of primary alignments: 2380751 [different]
Number of secondary alignments: 4397060 [different]
Number of bases in mapped sequences: 36042685390 [same]
Number of mapped bases: 35461625642 [different]
```

Thanks,
Alex

**Comment by SaurabhKalikar** [comment]
Hi Alex,

Glad to know that you could see the performance gains. 

Please note that AVX512 version of mm2-fast by default runs with a chaining parameter max-skip=infinity for higher chaining precision.
However, non-AVX512 code runs with default minimap2 settings, i.e., max-skip=25. That is why you see the difference in the output of your two runs. Your first output (with only lhash) should match with the original minimap2-v.2.18.

Example:
```
./minimap2-v2.18 -ax map-ont ref.fa reads.fastq >output1
./mm2-fast-lhash-only -ax map-ont ref.fa reads.fastq >output2
```
Here, output1 should be exactly same as output2.

In the case of mm2-fast with all AVX512 optimizations (your second output), the output should match minimap2-v2.18 with max-skip=infinity (using –max-cain-skip command-line flag). The details of --max-chain-skip flag can be found in minimap2’s man-page: https://lh3.github.io/minimap2/minimap2.html


Example: (with significantly large value for max-skip)

```
./minimap2-v2.18 -ax map-ont --max-chain-skip=10000000  ref.fa reads.fastq >output3
./mm2-fast-with-full-AVX512 -ax map-ont ref.fa reads.fastq >output4
```
Here, output3 should match exactly with output4.

Please note that, we are in the process of updating our README with the instructions to verify the correctness.

Could you also provide us your machine details and the compilation options you used for your two runs?

**Comment by ASLeonard** [comment]
Hi,

Glad it was something so simple, thanks for clarifying.

So far these have all been run on Xeon Gold 6140 nodes. Compiled with gcc8.2.0, everything was as described in your readme but with the flags `-flto -pipe`. The first binary was compiled on a head node without AVX512, so despite setting `march=skylake-avx512`, it didn't have the #defs needed and so only had the lhash enabled.

**Comment by cjain7** [comment]
Hi @ASLeonard , were you able to eventually verify correctness? We have now revised our README to better convey procedure for correctness check.

>I've seen between 3-7x speedup depending on the compiler options when mapping about 15x hifi data to a reference.

This is good. What HiFi sequencing (library, coverage, genome) and reference were you using? Your numbers appear to be better than what we observed in our benchmarks.

**Comment by ASLeonard** [comment]
Yes, the `--max-chain-skip=1000000` flag restored correctness. I noticed I was also defaulting to the tip version of minimap2, so after trying with v2.18, the same results are achieved. 

This is based on 20 Kb Hifi reads at about 15-20x coverage on a bovine assembly. 

Another thing I noticed is that the CPU time ratio is different to the wallclock time ratio. For example aligning two bovine assemblies together
```
AVX512 all:
CPU 4302s
Wall 1251s

minimap2:
CPU 9365s
Wall 2131s
```
So there is a ~2.2x speedup in CPU time, but only a 1.7x speedup in wallclock. It is pretty close, but the 3-7x I reported was based on CPU time, so maybe is closer to your estimates based on wallclock.
